game.py

```diff
@@ -0,0 +1,908 @@
+
+def get_human_move(game_state):
+
+    global rows
+    global columns
+       
+    board = game_state.get_board()
+    
+    while True:
+        try:
+            column = int(input("Enter column : ")) - 1
+            if column < 0 or column >= columns:
+                print('Column outside range. Try again')
+            else:
+                for row in range(rows):
+                    if board[row][column] == None:
+                        return row, column
+                print('Column full. Try again')                                
+        except:
+            print('Invalid input. Try again')
+                             
+def get_hint():
+    
+    hint = input("Get hint? (Y/N) : ")
+    if hint == 'Y' or hint == 'y':
+        return True
+    else:
+        return False
+
+def get_connect_number():
+    
+    while True:
+        try:
+            connect_number = int(input("Enter connect number : "))
+            if connect_number <= 0:
+                print('Connect number must be  positive: Try again')
+            else:
+                return connect_number
+        except:
+            print('Invalid input. Try again')
+
+def save_game(game_state):
+    
+    save = input("Save Game:? (Y/N) : ")
+    while save == 'Y' or save == 'y':
+        filename = input('Enter file name to save to : ')
+        try:
+            with open(filename, 'wb') as f:
+                pickle.dump(game_state, f)
+                break
+        except Exception as ex:
+            print('Error during saving game : ', ex)
+            save = input("Try again:? (Y/N) : ")
+         
+def restore_game():
+    
+    restore = input("Restore Game:? (Y/N) : ")
+    while restore == 'Y' or restore == 'y':
+        filename = input('Enter file name to restore : ')
+        try:
+            with open(filename, 'rb') as f:
+                return pickle.load(f)
+        except Exception as ex:
+            print('Error during restoring game : ', ex)
+            restore = input("Try again:? (Y/N) : ")
+
+def get_cyborg():
+    
+    cyborg = input("Cyborg? (Y/N) : ")
+    if cyborg == 'Y' or cyborg == 'y':
+        return True
+    else:
+        return False
+
+def get_score_find():
+    
+    global score_find
+   
+    while True:
+        try:
+            score_find = int(input("Enter score to find : "))
+            return score_find
+        except:
+            print('Invalid input. Try again')          
+def main():
+    
+    game_state = restore_game()
+    if game_state == None:
+        board = []
+        board_size = choose_board_size()
+        board = create_board(board_size)
+        connect_number = get_connect_number()
+        player_next = 1
+        number_of_moves = 0
+        game_state = GameState(board, player_next, connect_number, number_of_moves)
+
+    game_start_time = time.time()
+    global cyborg
+    global cyborg_look_forward
+    global cyborg_sort_depth
+    
+    global rows
+    global columns

```

Comment:
I think the proliferating use of globals is coming from a overdue need to refactor your class-hierarchy. All state should have a good home in a class. There are exceptions in python where you might want to set top-level variables that are imported into modules/sub-modules, but that isn't the case here.

The category of these variable are different, the `cyborg_x` ones get changed all the time (asking the user each time) and the others only get changed at each start of a game (row/column).

For `cyborg_x` variables, it seems that you have moved the player-experience to the idea of each move or game-situation being it's own thing, where you can choose your own distinct settings for how to play. I would name this and make it explicit in your code by creating a class, maybe called `GameSituationSettings`. This can sit on one of your other classes, and you can think more about how you'd ideally like to interact with it (e.g. I personally don't like this it seems I have to re-pick my settings each time rather than just tell it to remember my previous choices). This might not be the best thing to pick out, but I'd trying to pick out parts of state and behavior and extract them from the rest of your code. This method, for example is over 100 lines long and is unwieldy a this point.

For row/column, these below on a `Board` class in my opinion (maybe that concept is just subsumed into the `GameState` class, or maybe we have a separate `Board` class and have the `GameState` class hold a `Board` object in one of its fields).